[Pointer Events] Implement `getCoalescedEvents` API (macOS)

rr-codes opened the pull request:
#### bfd07b97a4631c62a55f8abfbc45f855d251ee3d
<pre>
[Pointer Events] Implement `getCoalescedEvents` API (macOS)
<a href="https://bugs.webkit.org/show_bug.cgi?id=276870">https://bugs.webkit.org/show_bug.cgi?id=276870</a>
<a href="https://rdar.apple.com/132027950">rdar://132027950</a>

Reviewed by Abrar Rahman Protyasha.

Implement the `getCoalescedEvents` function of the `PointerEvent` interface, as per the corresponding
UI events specification (<a href="https://w3c.github.io/pointerevents/#dom-pointerevent-getcoalescedevents).">https://w3c.github.io/pointerevents/#dom-pointerevent-getcoalescedevents).</a>

To facilitate this, the mousemouse events that were coalesced in the UI process are now stored and
propagated to the web process so that `getCoalescedEvents` returns them.

A future commit will enhance this implementation for iOS specifically by using the `coalescedTouchesForTouch`
UIKit method to gather more coalesced events.

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/coalesced_events_attributes.https_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/coalesced_events_attributes.https_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/coalesced_events_attributes_under_load.https_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/coalesced_events_attributes_under_load.https_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/extension/pointerevent_constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/idlharness.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_constructor.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerlock/pointerevent_getCoalescedEvents_when_pointerlocked.https-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/DragEvent.cpp:
(WebCore::DragEvent::DragEvent):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::dispatchMouseEvent):
(WebCore::Element::dispatchMouseForceWillBegin):
* Source/WebCore/dom/MouseEvent.cpp:
(WebCore::MouseEvent::create):
(WebCore::MouseEvent::MouseEvent):
* Source/WebCore/dom/MouseEvent.h:
(WebCore::MouseEvent::coalescedEvents const):
* Source/WebCore/dom/MouseRelatedEvent.h:
* Source/WebCore/dom/PointerEvent.cpp:
(WebCore::createCoalescedPointerEvents):
(WebCore::m_coalescedEvents):
(WebCore::PointerEvent::getCoalescedEvents const):
(WebCore::PointerEvent::receivedTarget):
(WebCore::PointerEvent::getCoalescedEvents): Deleted.
* Source/WebCore/dom/PointerEvent.h:
* Source/WebCore/dom/SimulatedClick.cpp:
* Source/WebCore/dom/WheelEvent.cpp:
(WebCore::WheelEvent::WheelEvent):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleMouseUp):
(WebCore::EventHandler::handleMouseMoveEvent):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/PointerCaptureController.h:
* Source/WebCore/platform/PlatformMouseEvent.h:
(WebCore::PlatformMouseEvent::coalescedEvents const):
* Source/WebKit/Shared/NativeWebMouseEvent.h:
* Source/WebKit/Shared/WebEvent.serialization.in:
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformMouseEvent::WebKit2PlatformMouseEvent):
* Source/WebKit/Shared/WebMouseEvent.cpp:
(WebKit::WebMouseEvent::WebMouseEvent):
* Source/WebKit/Shared/WebMouseEvent.h:
(WebKit::WebMouseEvent::WebMouseEvent):
(WebKit::WebMouseEvent::setCoalescedEvents):
(WebKit::WebMouseEvent::coalescedEvents const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::removeOldRedundantEvent):
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::handleMouseEvent):
(WebKit::WebPageProxy::processNextQueuedMouseEvent):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::navigateToURL):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::handleMouseEvent):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::navigateToPDFLinkWithSimulatedClick):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::inspectorNodeSearchMovedToPosition):
* Source/WebKitLegacy/mac/WebView/WebPDFView.mm:
(-[WebPDFView PDFViewWillClickOnLink:withURL:]):

Canonical link: <a href="https://commits.webkit.org/281387@main">https://commits.webkit.org/281387@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68bf003846481d80a94bc2110ef2beb8a358d37b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59642 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38987 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/12167 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63558 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10166 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61771 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46640 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/10320 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48384 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7117 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61672 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36387 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51629 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29222 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33090 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8877 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9089 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/52736 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55040 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9155 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/65289 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/58887 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3570 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9047 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55728 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3581 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51618 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55858 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2963 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/80644 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8927 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34801 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/13990 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35884 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36970 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35629 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -2831,17 +2831,17 @@ GenericCueAPIEnabled:
 
 GetCoalescedEventsEnabled:
   type: bool
-  status: unstable
```
change this to testable

rr-codes commented on Source/WebCore/dom/PointerEvent.cpp:
```diff
@@ -102,16 +102,55 @@ PointerEvent::PointerEvent(const AtomString& type, Init&& initializer)
+    //
+    //  Empty coalesced events list and predicted events list of their own.
+
+    MonotonicTime previousEventTimestamp;
```
this invariant assertion should probably only be executed in debug mode

aprotyas replied:
Do you mean about the timestamp monotonicity? Also, yeah I agree, maybe just wrap the whole evaluation behind an assert to ensure we don't even have to do the looping at runtime.

```cpp
auto coalescedEventSequenceTimestampsAreMonotonicallyIncreasing = [](const Vector& events) {
    return std::ranges::is_sorted(events, [](const auto& previousEvent, const auto& newEvent) {
        return previousEvent->timeStamp() <= newEvent->timeStamp();
    });
};

ASSERT(coalescedEventSequenceTimestampsAreMonotonicallyIncreasing(result));
```

rr-codes commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3566,11 +3570,11 @@ void WebPageProxy::handleMouseEvent(const NativeWebMouseEvent& event)
+        internals().coalescedMouseEvents.append(*removedEvent);
 
-    UNUSED_PARAM(didRemoveEvent);
-    LOG_WITH_STREAM(MouseHandling, stream << "UIProcess: " << (didRemoveEvent ? "replaced" : "enqueued") << " mouse event " << event.type() << " (queue size " << internals().mouseEventQueue.size() << ")");
```
put this log back

rr-codes commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3630,8 +3634,12 @@ void WebPageProxy::processNextQueuedMouseEvent()
         sandboxExtensions = SandboxExtension::createHandlesForMachLookup({ "com.apple.iconservices"_s, "com.apple.iconservices.store"_s }, process->auditToken(), SandboxExtension::MachBootstrapOptions::EnableMachBootstrap);
 #endif
 
-    LOG_WITH_STREAM(MouseHandling, stream << "UIProcess: sent mouse event " << eventType << " (queue size " << internals().mouseEventQueue.size() << ")");
```
put this log back

aprotyas commented on Source/WebCore/dom/Element.h:
```diff
@@ -686,7 +687,7 @@ class Element : public ContainerNode {
     enum class EventIsDispatched : bool { No, Yes };
     using DispatchMouseEventResult = std::pair<EventIsDispatched, EventIsDefaultPrevented>;
-    DispatchMouseEventResult dispatchMouseEvent(const PlatformMouseEvent& platformEvent, const AtomString& eventType, int clickCount = 0, Element* relatedTarget = nullptr, IsSyntheticClick = IsSyntheticClick::No);
+    DispatchMouseEventResult dispatchMouseEvent(const CoalescedPlatformMouseEvent& platformEvent, const AtomString& eventType, int clickCount = 0, Element* relatedTarget = nullptr, IsSyntheticClick = IsSyntheticClick::No);
```
Nit: Is the `platformEvent` parameter name needed?

aprotyas commented on Source/WebCore/dom/PointerEvent.cpp:
```diff
@@ -102,16 +102,55 @@ PointerEvent::PointerEvent(const AtomString& type, Init&& initializer)
+    //  Set the event's pointerId, pointerType, isPrimary and isTrusted to match the respective properties of the "parent" pointer event.
+    //  Set the event's cancelable and bubbles to false (as these events will never be dispatched in isolation).
+    //  Set the event's coalesced events list and predicted events list to an empty list.
+    //  Initialize all other attributes to default PointerEvent values.
```
It might be better to just link the appropriate standards text here instead of spelling out all the steps/constraints?

Maybe just a comment like this at the top of the function:

```
// Populated in accordance with https://www.w3.org/TR/pointerevents3/#populating-and-maintaining-the-coalesced-and-predicted-events-lists
```

aprotyas commented on Source/WebCore/dom/PointerEvent.cpp:
```diff
@@ -128,9 +167,22 @@ PointerEvent::PointerEvent(const AtomString& type, PointerID pointerId, const St
+    // When a trusted PointerEvent's target is changed, user agents SHOULD, for each event in the
+    // coalesced events list and predicted events list:
+    //
+    //  Set the event's target to match the target of the "parent" pointer event.
```
I don't think this comment is necessary. This is just the standards text and doesn't explain why we did this, and the code is super readable here.

aprotyas reviewed (commented):
I'm in general agreement with the population (and maintenance) of coalesced pointer events, but a bit worried about the mechanical parts of this change.

We seem to be proliferating the `Coalesced[Platform]MouseEvent` type (and referencing `Coalesced[Platform]MouseEvent.parent`) across our event handling and creation code paths. This adds an extra complexity factor in some (already) dense code paths.

I can't suggest a different approach off the top of my head, but I'm wondering if we can further abstract this concept of an event possibly being coalesced.

---

Is there a plan for the legit failure in `imported/w3c/web-platform-tests/pointerevents/coalesced_events_attributes_under_load.https`? Were you able to diagnose how severe this failure is (if it has any user impact?)

If we commit the failing expectations, which are no longer because of the missing API, let's track the failure through Bugzilla.

aprotyas commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3630,8 +3632,18 @@ void WebPageProxy::processNextQueuedMouseEvent()
+        eventWithCoalescedEvents.setCoalescedEvents(internals().coalescedMouseEvents);
+    }
+
+    LOG_WITH_STREAM(MouseHandling, stream << "UIProcess: sent mouse event " << eventType << " (queue size " << internals().mouseEventQueue.size() << << ", coalesced events size " << internals().coalescedMouseEvents << ")");
```
Debug build is sad because of this

aprotyas commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -487,15 +487,21 @@ Element::DispatchMouseEventResult Element::dispatchMouseEvent(const PlatformMous
+        childMouseEvents.append(WTFMove(childMouseEvent));
+    }
+
+    Ref parentMouseEvent = MouseEvent::create(eventType, document().windowProxy(), platformEvent, childMouseEvents, detail, relatedTarget);
```
I think maintaining `coalesced vs primary/main` throughout rather than introducing `child vs parent` would be easier to follow. Would a child event necessarily mean it's coalesced?

aprotyas commented on Source/WebKit/Shared/WebEventConversion.cpp:
```diff
@@ -117,6 +117,9 @@ class WebKit2PlatformMouseEvent : public WebCore::PlatformMouseEvent {
         m_unadjustedMovementDelta = webEvent.unadjustedMovementDelta();
         m_globalPosition = webEvent.globalPosition();
         m_clickCount = webEvent.clickCount();
+        m_coalescedEvents = WTF::map(webEvent.coalescedEvents(), [&](const auto& event) {
```
You can get away without capturing anything, I believe

rr-codes committed:
[Pointer Events] Implement `getCoalescedEvents` API (macOS)
https://bugs.webkit.org/show_bug.cgi?id=276870
rdar://132027950

Reviewed by Abrar Rahman Protyasha.

Implement the `getCoalescedEvents` function of the `PointerEvent` interface, as per the corresponding
UI events specification (https://w3c.github.io/pointerevents/#dom-pointerevent-getcoalescedevents).

To facilitate this, the mousemouse events that were coalesced in the UI process are now stored and
propagated to the web process so that `getCoalescedEvents` returns them.

A future commit will enhance this implementation for iOS specifically by using the `coalescedTouchesForTouch`
UIKit method to gather more coalesced events.

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/coalesced_events_attributes.https_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/coalesced_events_attributes.https_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/coalesced_events_attributes_under_load.https_mouse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/coalesced_events_attributes_under_load.https_pen-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/extension/pointerevent_constructor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/idlharness.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_constructor.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerlock/pointerevent_getCoalescedEvents_when_pointerlocked.https-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/DragEvent.cpp:
(WebCore::DragEvent::DragEvent):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::dispatchMouseEvent):
(WebCore::Element::dispatchMouseForceWillBegin):
* Source/WebCore/dom/MouseEvent.cpp:
(WebCore::MouseEvent::create):
(WebCore::MouseEvent::MouseEvent):
* Source/WebCore/dom/MouseEvent.h:
(WebCore::MouseEvent::coalescedEvents const):
* Source/WebCore/dom/MouseRelatedEvent.h:
* Source/WebCore/dom/PointerEvent.cpp:
(WebCore::createCoalescedPointerEvents):
(WebCore::m_coalescedEvents):
(WebCore::PointerEvent::getCoalescedEvents const):
(WebCore::PointerEvent::receivedTarget):
(WebCore::PointerEvent::getCoalescedEvents): Deleted.
* Source/WebCore/dom/PointerEvent.h:
* Source/WebCore/dom/SimulatedClick.cpp:
* Source/WebCore/dom/WheelEvent.cpp:
(WebCore::WheelEvent::WheelEvent):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::handleMouseUp):
(WebCore::EventHandler::handleMouseMoveEvent):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/PointerCaptureController.h:
* Source/WebCore/platform/PlatformMouseEvent.h:
(WebCore::PlatformMouseEvent::coalescedEvents const):
* Source/WebKit/Shared/NativeWebMouseEvent.h:
* Source/WebKit/Shared/WebEvent.serialization.in:
* Source/WebKit/Shared/WebEventConversion.cpp:
(WebKit::WebKit2PlatformMouseEvent::WebKit2PlatformMouseEvent):
* Source/WebKit/Shared/WebMouseEvent.cpp:
(WebKit::WebMouseEvent::WebMouseEvent):
* Source/WebKit/Shared/WebMouseEvent.h:
(WebKit::WebMouseEvent::WebMouseEvent):
(WebKit::WebMouseEvent::setCoalescedEvents):
(WebKit::WebMouseEvent::coalescedEvents const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::removeOldRedundantEvent):
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::handleMouseEvent):
(WebKit::WebPageProxy::processNextQueuedMouseEvent):
(WebKit::WebPageProxy::resetStateAfterProcessExited):
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::navigateToURL):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::handleMouseEvent):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::navigateToPDFLinkWithSimulatedClick):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::inspectorNodeSearchMovedToPosition):
* Source/WebKitLegacy/mac/WebView/WebPDFView.mm:
(-[WebPDFView PDFViewWillClickOnLink:withURL:]):

Canonical link: https://commits.webkit.org/281387@main
